### PR TITLE
Make it possible for from_diff to support custom types (issue #107)

### PR DESCRIPTION
**Pull request by clj**
This makes it possible to pass a custom `dumps` implementation to `from_diff` to support types not natively supported by Python's json module.

I have not dealt with `from_string` and `to_string` which also use the json module. The easy solution here would be to deprecate them and rely on people using whichever implementation of json dumping and loading they prefer.

### CODE REVIEW

**Comment by stefankoegl** [comment]
Thanks for the pull request - looks great!

Could you add a simple example to the [tutorial](https://github.com/stefankoegl/python-json-patch/blob/master/doc/tutorial.rst)? Thanks :)

**Comment by clj** [comment]
I will do. 

Do you have any opinion on what to do about `from_string` and `to_string`?

**Comment by stefankoegl** [comment]
I see no reason to deprecate those two. We could also add an optional parameter to those two functions.

**Comment by clj** [comment]
@stefankoegl sorry about the delay in getting the PR updated, real life intervened.

I have added some documentation as requested, and used `Decimal` and the simplejson package as an example, as I imagine this being the primary use case for this change.

I have also made it possible to subclass `JsonPatch` to change its behaviour so that it is not necessary to pass dumps/loads arguments every time the individual methods are called.

Let me know if you have any comments and I'll get things turned around quicker this time :)

File: jsonpatch.py [hunk 1]
```diff
@@ -637,7 +649,8 @@ def apply(self, obj):
 
 class DiffBuilder(object):
 
-    def __init__(self):
+    def __init__(self, dumps):
```
Comment (stefankoegl) [inline jsonpatch.py]: Could we make this an optional argument?

Comment (clj) [reply jsonpatch.py]: Done. Defaults to `json.dumps`

**Review by Alanscut — APPROVED** [review]
@stefankoegl should we merge this PR?

**Review by stefankoegl — APPROVED** [review]

**Comment by stefankoegl** [comment]
Merged, thanks :)

**Pull request merged**